internal/network/none_platform_node_ips_allocation.go

```diff
@@ -0,0 +1,312 @@
+	// that node-ip allocation on these networks should be avoided
+	prohibitedNetworks := lo.Filter(hostCandidates.hostNetworks,
+		func(hostNetwork *hostNetwork, _ int) bool {
+			return hostNetwork.ipNet.String() != eligibleNodeIp.ipNet.String()
+		})
+	return &collisionDetector{
+		assignedNodeIps: []*nodeIpCandidate{{
+			hostId: hostCandidates.hostId,
+			ip:     eligibleNodeIp.nodeIp,
+		}},
+		prohibitedNetworks: prohibitedNetworks,
+	}
+}
+
+func (c *collisionDetector) detectIpNetworkCollision(nodeIpCandidate *nodeIpCandidate, prohibitedNetworks []*hostNetwork) []error {
+	var errs []error
+	for _, prohibitedNetwork := range prohibitedNetworks {
+		if prohibitedNetwork.ipNet.Contains(nodeIpCandidate.ip) {
+			errs = append(errs,
+				errors.Errorf("IP assignment %s in host %s collides with prohibited network %s in host %s",
+					nodeIpCandidate.ip, nodeIpCandidate.hostId, prohibitedNetwork.ipNet.String(), prohibitedNetwork.hostId))
+		}
+	}
+	return errs
+}
+
+func (c *collisionDetector) detectIpsNetworksCollisions(assignedNodeIps []*nodeIpCandidate, prohibitedNetworks []*hostNetwork) []error {
+	var errs []error
+	for _, assignment := range assignedNodeIps {
+		errs = append(errs, c.detectIpNetworkCollision(assignment, prohibitedNetworks)...)
+	}
+	return errs
+}
+
+func (c *collisionDetector) detectCollisions(other *collisionDetector) []error {
+	var errs []error
+	errs = append(append(errs,
+		c.detectIpsNetworksCollisions(c.assignedNodeIps, other.prohibitedNetworks)...),
+		c.detectIpsNetworksCollisions(other.assignedNodeIps, c.prohibitedNetworks)...)
+	return errs
+}
+
+func (c *collisionDetector) merge(other *collisionDetector) *collisionDetector {
+	return &collisionDetector{
+		assignedNodeIps:    append(c.assignedNodeIps, other.assignedNodeIps...),
+		prohibitedNetworks: append(c.prohibitedNetworks, other.prohibitedNetworks...),
+	}
+}
+
+type nodeIpAllocator struct {
+	candidates []*hostCandidates
+	clusterId  strfmt.UUID
+	log        logrus.FieldLogger
+}
+
+func (n *nodeIpAllocator) sort(allocations []*allocationSolution) {
+	sort.SliceStable(allocations, func(i, j int) bool {
+		return allocations[i].routeMetric < allocations[j].routeMetric
+	})
+}
+
+// Allocate node IPs.  Since the node IPs are dependent on each other, then all of them must be allocated
+// through this allocation procedure.  It tries to find an allocation when no conflict exists.
+func (n *nodeIpAllocator) allocate(index int, parentContext *collisionDetector) (map[strfmt.UUID]*NodeIpAllocation, []error) {
+	if index == len(n.candidates) {
+		return make(map[strfmt.UUID]*NodeIpAllocation), nil
+	}
+	var (
+		errs                []error
+		allocationSolutions []*allocationSolution
+		h                   = n.candidates[index]
+	)
+	for i := range h.eligibleNodeIps {
+		c := h.eligibleNodeIps[i]
+		localCollisionContext := newCollisionDetectorForNodeIp(h, c)
+		if collisionErrs := localCollisionContext.detectCollisions(parentContext); len(collisionErrs) > 0 {
```

Comment:
Add logging of the localCollisionContext of each host for debuging

Reply:
Bare in mind that it can be multiple times for the same host